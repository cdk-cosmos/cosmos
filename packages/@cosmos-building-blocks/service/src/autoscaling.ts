import { Construct, Lazy, Fn } from '@aws-cdk/core';
import { InstanceType, InstanceClass, InstanceSize, CfnLaunchTemplate } from '@aws-cdk/aws-ec2';
import { CfnInstanceProfile } from '@aws-cdk/aws-iam';
import {
  CfnAutoScalingGroup,
  AutoScalingGroup,
  AutoScalingGroupProps,
  Monitoring,
  BlockDevice,
  EbsDeviceVolumeType,
  BlockDeviceVolume,
} from '@aws-cdk/aws-autoscaling';

/**
 * Properties for AutoScalingGroupLT
 */
export interface AutoScalingGroupLTProps extends AutoScalingGroupProps {
  /**
   * Set this to `true` if you want to use Instance Template instead of Launch Config.
   * You can fine tune your default LaunchTemplate with `instanceType`, `launchTemplateOverrides` and `instancesDistribution`
   * @default - false
   */
  readonly useInstanceTemplate?: boolean;
  /**
   * Type of instance to launch. if used with `useInstanceTemplate = true`, it will be applied to LaunchTemplate.
   * Make sure to check and tweak `launchTemplateOverrides` and `instancesDistribution` as the default values set for those
   * properties will overwrite this InstanceType
   */
  readonly instanceType: InstanceType;
  /**
   * Can only be used along with `useInstanceTemplate = true`.
   * Array of LaunchTemplateOverridesProperty which is a subproperty of LaunchTemplate that describes an override for a launch template.
   * Currently, the only supported override is instance type
   * https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-autoscaling-autoscalinggroup-launchtemplateoverrides.html
   * @default
   * [{ instanceType: "t3.medium" }, { instanceType: "t3.small" }]
   *
   * This will use any Spot instances available, "t3.medium" or "t3.small"
   * based on `spotAllocationStrategy` set in `instancesDistribution`
   */
  readonly launchTemplateOverrides?: Array<CfnAutoScalingGroup.LaunchTemplateOverridesProperty>;
  /**
   * Can only be used along with `useInstanceTemplate = true`.
   * Specifies the distribution of On-Demand Instances and Spot Instances,
   * the maximum price to pay for Spot Instances, and how the Auto Scaling group allocates instance types
   * to fulfill On-Demand and Spot capacity.
   * https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-autoscaling-autoscalinggroup-instancesdistribution.html
   * @default - Uses all Spot instances of types specified in property `launchTemplateOverrides`
   *  {
   *    onDemandAllocationStrategy: "prioritized",
   *    onDemandBaseCapacity: 0,
   *    onDemandPercentageAboveBaseCapacity: 0,
   *    spotAllocationStrategy: "capacity-optimized",
   *    spotMaxPrice: props.spotPrice,
   *  }
   */
  readonly instancesDistribution?: Partial<CfnAutoScalingGroup.InstancesDistributionProperty>;
}

/**
 * Derived from Base class AutoScalingGroup with additional functionalities of using Launch Template (LT).
 *
 * Base AutoScalingGroup class uses Launch Config and have no ability to use more advanced Launch Templates
 * and MixedInstancesPolicies which gives ability to use Spot fleet and combination of spot and on-demand
 *
 * This construct uses default spot fleet template settings with just one property `useInstanceTemplate = true`
 *
 * With `launchTemplateOverrides` and `instancesDistribution`, it gives additional fine tuning.
 */
export class AutoScalingGroupLT extends AutoScalingGroup {
  constructor(scope: Construct, id: string, props: AutoScalingGroupLTProps) {
    super(scope, id, props);

    // Check if useInstanceTemplate is set to true. If not, do nothing and rely on Base class to
    // to create AutoScalingGroup using LaunchConfig
    const useInstanceTemplate = props.useInstanceTemplate !== undefined ? props.useInstanceTemplate === true : false;
    if (useInstanceTemplate == true) {
      // Prepare props required for LaunchTemplate
      const instanceType =
        props.instanceType !== undefined
          ? props.instanceType.toString()
          : InstanceType.of(InstanceClass.T3, InstanceSize.SMALL).toString();

      const imageConfig = props.machineImage.getImage(this);
      const userDataToken = Lazy.stringValue({
        produce: () => Fn.base64(this.userData.render()),
      });
      const securityGroupIds = this.connections.securityGroups.map(sg => sg.securityGroupId);
      const instanceMonitoringMode =
        props.instanceMonitoring !== undefined ? props.instanceMonitoring === Monitoring.DETAILED : undefined;
      const monitoring = {
        enabled: instanceMonitoringMode == undefined ? true : instanceMonitoringMode,
      };

      const iamProfile = this.node.tryFindChild('InstanceProfile') as CfnInstanceProfile;

      // Create LaunchTemplate

      const lTemplate = new CfnLaunchTemplate(this, 'LaunchTemplate', {
        launchTemplateData: {
          imageId: imageConfig.imageId,
          securityGroupIds,
          userData: userDataToken,
          iamInstanceProfile: { arn: iamProfile.attrArn },
          keyName: props.keyName,
          monitoring,
          instanceType: instanceType,
          blockDeviceMappings:
            props.blockDevices !== undefined ? synthesizeBlockDeviceMappings(this, props.blockDevices) : undefined,
        },
      });
      lTemplate.node.addDependency(this.role);

      // Prepare props for ASG with MixedInstancesPolicy

      const overrides =
        props.launchTemplateOverrides == undefined
          ? [
              {
                instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MEDIUM).toString(),
              },
              {
                instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.SMALL).toString(),
              },
            ]
          : props.launchTemplateOverrides;

      const instancesDistribution = {
        onDemandAllocationStrategy: 'prioritized',
        onDemandBaseCapacity: 0,
        onDemandPercentageAboveBaseCapacity: 0,
        spotAllocationStrategy: 'capacity-optimized',
        spotMaxPrice: props.spotPrice,
        ...props.instancesDistribution,
      };

      // Mixed instance Policy
      const mixedInstancesPolicy: CfnAutoScalingGroup.MixedInstancesPolicyProperty = {
        launchTemplate: {
          launchTemplateSpecification: {
            version: lTemplate.attrLatestVersionNumber,
            launchTemplateId: lTemplate.ref,
          },
          overrides,
        },
        instancesDistribution,
      };

      // Remove LaunchConfig from the resulting CloudFormation template
      if (!this.node.tryRemoveChild('LaunchConfig')) throw new Error('Launch config not found');
      const asg = this.node.tryFindChild('ASG') as CfnAutoScalingGroup;
      asg.addPropertyDeletionOverride('LaunchConfigurationName');
      // Inject mixedInstancesPolicy to use LaunchTemplate
      asg.mixedInstancesPolicy = mixedInstancesPolicy;
    }
  }
}
function synthesizeBlockDeviceMappings(
  construct: Construct,
  blockDevices: BlockDevice[]
): CfnLaunchTemplate.BlockDeviceMappingProperty[] {
  return blockDevices.map<CfnLaunchTemplate.BlockDeviceMappingProperty>(({ deviceName, volume, mappingEnabled }) => {
    const { virtualName, ebsDevice: ebs } = volume;

    if (volume === BlockDeviceVolume.noDevice() || mappingEnabled === false) {
      return {
        deviceName,
        noDevice: 'true',
      };
    }

    if (ebs) {
      const { iops, volumeType } = ebs;

      if (!iops) {
        if (volumeType === EbsDeviceVolumeType.IO1) {
          throw new Error('iops property is required with volumeType: EbsDeviceVolumeType.IO1');
        }
      } else if (volumeType !== EbsDeviceVolumeType.IO1) {
        construct.node.addWarning('iops will be ignored without volumeType: EbsDeviceVolumeType.IO1');
      }
    }

    return {
      deviceName,
      ebs,
      virtualName,
    };
  });
}
