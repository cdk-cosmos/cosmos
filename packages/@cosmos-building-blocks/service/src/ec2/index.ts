export {
  CloudWatchAgent,
  CloudWatchAgentProps,
  CloudWatchMetricBuilder,
  CloudWatchMetric,
} from './ec2-cloud-watch-agent';
