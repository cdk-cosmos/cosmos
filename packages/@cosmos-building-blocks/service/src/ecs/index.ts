export { EcsService, EcsServiceProps } from './ecs-service';
