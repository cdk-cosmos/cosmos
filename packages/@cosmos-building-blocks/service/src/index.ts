export { EcsService, EcsServiceProps } from './ecs';
