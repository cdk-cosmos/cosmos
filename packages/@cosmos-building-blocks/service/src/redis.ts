import { Construct, IConstruct, Stack } from '@aws-cdk/core';
import { CfnSubnetGroup, CfnReplicationGroup, CfnReplicationGroupProps } from '@aws-cdk/aws-elasticache';
import {
  IVpc,
  ISubnet,
  IConnectable,
  ISecurityGroup,
  SecurityGroup,
  Port,
  SubnetSelection,
  Connections,
  Peer,
} from '@aws-cdk/aws-ec2';

export interface IRedis extends IConstruct {
  /**
   * Security Group assigned to Redis Cluster
   */
  readonly redisSecurityGroups?: ISecurityGroup[];
  /**
   * Redis Protocol of the Primary Address connection (redis[s])
   */
  readonly redisProtocol: string;
  /**
   * Redis endpoint of the Primary Address
   */
  readonly redisEndpoint: string;
  /**
   * Redis port of the Primary Address
   */
  readonly redisPort: string;
  /**
   * Redis URI `{protocol}://{endpoint}:{port} of the Primary Address
   */
  readonly redisUri: string;
}

/**
 * Redis Cluster props. for the optional props not provided, following defaults will be used:
 * Refer https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticache-replicationgroup.html for more
 * details on each prop
 *
 * @default
 *  replicationGroupDescription: 'Redis cluster', // Mandatory Description
 *  atRestEncryptionEnabled: true, // Enable encryption at rest
 *  cacheNodeType: 'cache.t3.micro', // Instance type
 *  numCacheClusters: 1, // The number of clusters this replication group initially has
 *  transitEncryptionEnabled: true, // Enable transit encryption
 *  automaticFailoverEnabled: false, // Specifies whether a read-only replica is automatically promoted to read/write primary if the existing primary fails.
 */
export interface RedisProps extends Omit<Partial<CfnReplicationGroupProps>, 'securityGroupIds'> {
  /**
   * VPC to launch Redis cluster in
   */
  readonly vpc: IVpc;
  /**
   * List of Subnets to launch Redis cluster in
   */
  readonly subnets: SubnetSelection[];
  /**
   * `AWS::ElastiCache::ReplicationGroup.SecurityGroupIds`.
   *
   * @default - Create a new Security Group for redis.
   */
  readonly securityGroups?: ISecurityGroup[];
  /**
   * Allowed ingress connections to Redis SecurityGroup
   *
   * @default - Allowed from anywhere.
   */
  readonly connectionsAllowedFrom?: string | IConnectable;
  /**
   * Enable sensible defaults for basic high availability
   *
   * @default - false.
   */
  readonly highAvailabilityMode?: boolean;
}

export class Redis extends Construct implements IRedis {
  /**
   * Actual Redis Cluster
   */
  readonly redisReplicationGroup: CfnReplicationGroup;
  /**
   * Subnet Group created by construct using Subnets provided in props
   */
  readonly cacheSubnetGroup: CfnSubnetGroup;
  /**
   * Security Group assigned to Redis Cluster
   */
  readonly redisSecurityGroups: ISecurityGroup[];
  /**
   * Redis Protocol of the Primary Address connection (redis[s])
   */
  readonly redisProtocol: string;
  /**
   * Redis endpoint of the Primary Address
   */
  readonly redisEndpoint: string;
  /**
   * Redis port of the Primary Address
   */
  readonly redisPort: string;
  /**
   * Redis URI `{protocol}://{endpoint}:{port} of the Primary Address
   */
  readonly redisUri: string;

  constructor(scope: Construct, id: string, props: RedisProps) {
    super(scope, id);

    const { vpc, subnets, connectionsAllowedFrom, port = 6379, securityGroups, highAvailabilityMode = false } = props;

    const subnetSelections = subnets
      .map(selection => vpc.selectSubnets(selection))
      .reduce<ISubnet[]>((subnets, selection) => {
        subnets.push(...selection.subnets.filter(x => !subnets.includes(x)));
        return subnets;
      }, []);

    // Create CacheSubnet group
    this.cacheSubnetGroup = new CfnSubnetGroup(this, 'RedisSubnetGroup', {
      description: `Redis Subnet group for ${this.node.path}`,
      subnetIds: subnetSelections.map(x => x.subnetId),
    });

    if (securityGroups && securityGroups.length) {
      this.redisSecurityGroups = securityGroups;
    } else {
      // Create SecurityGroup
      this.redisSecurityGroups = [
        new SecurityGroup(this, 'RedisSecurityGroup', {
          vpc: vpc,
          description: `Redis Security Group for ${this.node.path}`,
          allowAllOutbound: false,
        }),
      ];

      // Check if Security Ingress is required to be locked down
      if (connectionsAllowedFrom == undefined) {
        this.redisSecurityGroups[0].connections.allowFromAnyIpv4(Port.tcp(port));
      } else {
        this.redisSecurityGroups[0].connections.allowFrom(
          typeof connectionsAllowedFrom === 'string'
            ? new Connections({
                peer: Peer.ipv4(connectionsAllowedFrom),
              })
            : connectionsAllowedFrom,
          Port.tcp(port)
        );
      }
    }

    const modeProps: Partial<CfnReplicationGroupProps> = highAvailabilityMode
      ? {
          numNodeGroups: 1,
          replicasPerNodeGroup: Math.max(1, vpc.availabilityZones.length - 1),
          multiAzEnabled: true,
          automaticFailoverEnabled: true,
        }
      : {
          numNodeGroups: 1,
          replicasPerNodeGroup: 0,
          multiAzEnabled: false,
          automaticFailoverEnabled: false,
        };

    // Create the actual redis cluster
    this.redisReplicationGroup = new CfnReplicationGroup(this, 'Redis', {
      replicationGroupDescription: `Redis cluster for ${this.node.path}`,
      cacheNodeType: 'cache.t3.micro',
      atRestEncryptionEnabled: true,
      transitEncryptionEnabled: true,
      ...modeProps,
      ...props,
      cacheSubnetGroupName: this.cacheSubnetGroup.ref,
      securityGroupIds: this.redisSecurityGroups.map(x => x.securityGroupId),
      engine: 'redis',
    });

    this.redisProtocol = Stack.of(this).resolve(this.redisReplicationGroup.transitEncryptionEnabled)
      ? 'rediss'
      : 'redis';
    this.redisEndpoint = this.redisReplicationGroup.attrPrimaryEndPointAddress;
    this.redisPort = this.redisReplicationGroup.attrPrimaryEndPointPort;
    this.redisUri = `${this.redisProtocol}://${this.redisEndpoint}:${this.redisPort}`;
  }

  static fromRedisAttributes(scope: Construct, id: string, attrs: RedisAttributes): IRedis {
    return new ImportedRedis(scope, id, attrs);
  }
}

export interface RedisAttributes {
  /**
   * Security Group assigned to Redis Cluster
   */
  readonly redisSecurityGroups?: ISecurityGroup[];
  /**
   * Redis Protocol of the Primary Address connection (redis[s])
   */
  readonly redisProtocol: string;
  /**
   * Redis endpoint of the Primary Address
   */
  readonly redisEndpoint: string;
  /**
   * Redis port of the Primary Address
   */
  readonly redisPort: string;
}

class ImportedRedis extends Construct implements IRedis {
  /**
   * Security Group assigned to Redis Cluster
   */
  readonly redisSecurityGroups?: ISecurityGroup[];
  /**
   * Redis Protocol of the Primary Address connection (redis[s])
   */
  readonly redisProtocol: string;
  /**
   * Redis endpoint of the Primary Address
   */
  readonly redisEndpoint: string;
  /**
   * Redis port of the Primary Address
   */
  readonly redisPort: string;
  /**
   * Redis URI `{protocol}://{endpoint}:{port} of the Primary Address
   */
  readonly redisUri: string;

  constructor(scope: Construct, id: string, props: RedisAttributes) {
    super(scope, id);
    const { redisSecurityGroups, redisEndpoint, redisPort, redisProtocol } = props;

    this.redisSecurityGroups = redisSecurityGroups;
    this.redisProtocol = redisProtocol;
    this.redisEndpoint = redisEndpoint;
    this.redisPort = redisPort;
    this.redisUri = `${this.redisProtocol}://${this.redisEndpoint}:${this.redisPort}`;
  }
}
