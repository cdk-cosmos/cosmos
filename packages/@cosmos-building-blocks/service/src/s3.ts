import { Construct } from '@aws-cdk/core';
import { Bucket, BucketEncryption, BucketProps, BucketAccessControl } from '@aws-cdk/aws-s3';
import { PolicyStatement, Effect, AnyPrincipal } from '@aws-cdk/aws-iam';

export class SecureBucket extends Bucket {
  constructor(scope: Construct, id: string, props?: BucketProps) {
    super(scope, id, {
      encryption: BucketEncryption.S3_MANAGED,
      accessControl: BucketAccessControl.LOG_DELIVERY_WRITE,
      serverAccessLogsPrefix: 'BucketAccessLogs/',
      ...props,
    });
    const policyStatement = new PolicyStatement({
      actions: ['s3:*'],
      principals: [new AnyPrincipal()],
      resources: [`${this.bucketArn}/*`],
      effect: Effect.DENY,
      sid: 'ForceSSLOnly',
    });
    policyStatement.addCondition('Bool', { 'aws:SecureTransport': false });
    super.addToResourcePolicy(policyStatement);
  }
}
