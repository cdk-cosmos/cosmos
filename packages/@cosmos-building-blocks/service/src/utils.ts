export const getRoutingPriority = (props: { hostHeader?: string; pathPattern?: string }): number => {
  const path = getStringValue(props.pathPattern);
  const host = getStringValue(props.hostHeader);
  const count = 2;
  const total = path + host;
  const inMin = 0 * count;
  const inMax = 128 * charMapping.length * count;
  const outMin = 1000; // Offset
  const outMax = Math.min(inMax + outMin, 50000); // only use and many as needed ?

  // Since lower numbers are more important in Listener Rules for ALB, then flip out min and max
  return mapNumber(total, inMin, inMax, outMax, outMin);
};

const mapNumber = (value: number, inMin: number, inMax: number, outMin: number, outMax: number): number => {
  return Math.floor(((value - inMin) / (inMax - inMin)) * (outMax - outMin) + outMin);
};

const getStringValue = (str?: string): number => {
  if (str === undefined) return 0;
  let total = 0;
  for (let i = 0; i <= str.length; i++) {
    total += charMapping.indexOf(str.charAt(i)) + 1;
  }
  return total;
};

const charMapping: string[] = [
  '*',
  '?',
  '&',
  '_',
  '@',
  '+',
  "'",
  '"',
  '~',
  '$',
  '.',
  ':',
  '-',
  'a',
  'b',
  'c',
  'd',
  'e',
  'f',
  'g',
  'h',
  'i',
  'j',
  'k',
  'l',
  'm',
  'n',
  'o',
  'p',
  'q',
  'r',
  's',
  't',
  'u',
  'v',
  'w',
  'x',
  'y',
  'z',
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z',
  '0',
  '1',
  '2',
  '3',
  '4',
  '5',
  '6',
  '7',
  '8',
  '9',
  '/',
];
