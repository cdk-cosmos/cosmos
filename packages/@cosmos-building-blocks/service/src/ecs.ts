import { Construct, Duration } from '@aws-cdk/core';
import {
  Ec2TaskDefinition,
  Ec2Service,
  Protocol,
  ContainerDefinitionOptions,
  PortMapping,
  Ec2ServiceProps,
  ICluster,
  LogDrivers,
  ContainerDefinition,
  ScalableTaskCount,
  CpuUtilizationScalingProps,
  RequestCountScalingProps,
  PlacementStrategy,
  BuiltInAttributes,
  Ec2TaskDefinitionProps,
} from '@aws-cdk/aws-ecs';
import {
  ApplicationTargetGroup,
  ApplicationProtocol,
  ApplicationListenerRule,
  IApplicationListener,
  ApplicationListenerRuleProps,
  ApplicationTargetGroupProps,
} from '@aws-cdk/aws-elasticloadbalancingv2';
import { IVpc } from '@aws-cdk/aws-ec2';
import { LogGroup, LogGroupProps } from '@aws-cdk/aws-logs';
import { EnableScalingProps } from '@aws-cdk/aws-applicationautoscaling';
import { getRoutingPriority } from './utils';

export interface EcsServiceProps {
  vpc: IVpc;
  cluster: ICluster;
  containerProps: ContainerDefinitionOptions & { port?: PortMapping };
  httpListener?: IApplicationListener;
  taskProps?: Partial<Ec2TaskDefinitionProps>;
  serviceProps?: Partial<Ec2ServiceProps>;
  targetGroupProps?: Partial<ApplicationTargetGroupProps>;
  routingProps?: Partial<ApplicationListenerRuleProps>;
  scalingProps?: EnableScalingProps;
  logProps?: Partial<LogGroupProps>;
}

export class EcsService extends Construct {
  readonly LogGroup: LogGroup;
  readonly TaskDefinition: Ec2TaskDefinition;
  readonly Container: ContainerDefinition;
  readonly Service: Ec2Service;
  readonly TargetGroup?: ApplicationTargetGroup;
  readonly ListenerRule?: ApplicationListenerRule;
  readonly Scaling?: ScalableTaskCount;

  constructor(scope: Construct, id: string, props: EcsServiceProps) {
    super(scope, id);

    const {
      vpc,
      cluster,
      httpListener,
      containerProps,
      taskProps,
      serviceProps,
      targetGroupProps,
      routingProps,
      scalingProps,
      logProps,
    } = props;

    this.LogGroup = new LogGroup(this, 'Logs', logProps);

    this.TaskDefinition = new Ec2TaskDefinition(this, 'Task', {
      ...taskProps,
    });

    this.Container = this.TaskDefinition.addContainer('Container', {
      memoryLimitMiB: 256,
      logging: LogDrivers.awsLogs({ logGroup: this.LogGroup, streamPrefix: `Container` }),
      ...containerProps,
    });

    this.Container.addPortMappings({
      containerPort: 80,
      protocol: Protocol.TCP,
      ...containerProps.port,
    });

    this.Service = new Ec2Service(this, 'Service', {
      desiredCount: 1,
      placementStrategies: [PlacementStrategy.spreadAcross(BuiltInAttributes.AVAILABILITY_ZONE)],
      ...serviceProps,
      taskDefinition: this.TaskDefinition,
      cluster: cluster,
    });

    if (routingProps) {
      if (!httpListener) throw new Error('To enable routing, Http Listener is required');

      this.TargetGroup = new ApplicationTargetGroup(this, 'ServiceTargetGroup', {
        protocol: ApplicationProtocol.HTTP,
        deregistrationDelay: Duration.seconds(0),
        ...targetGroupProps,
        vpc: vpc,
        targets: [
          this.Service.loadBalancerTarget({
            containerName: 'Container',
          }),
        ],
      });

      this.ListenerRule = new ApplicationListenerRule(this, 'ServiceRule', {
        priority: getRoutingPriority(routingProps),
        ...routingProps,
        listener: httpListener,
        targetGroups: [this.TargetGroup],
      });
    }

    if (scalingProps) {
      this.Scaling = this.Service.autoScaleTaskCount(scalingProps);
    }
  }

  addCpuAutoScaling(props: Partial<CpuUtilizationScalingProps>): void {
    if (!this.Scaling) throw new Error('Scaling needs to be enabled');
    this.Scaling.scaleOnCpuUtilization('CpuScaling', {
      ...props,
      targetUtilizationPercent: 50,
    });
  }

  addMemoryAutoScaling(props: Partial<RequestCountScalingProps>): void {
    if (!this.Scaling) throw new Error('Scaling needs to be enabled');
    this.Scaling.scaleOnMemoryUtilization('MemoryScaling', {
      ...props,
      targetUtilizationPercent: 50,
    });
  }

  addRequestAutoScaling(props: Partial<RequestCountScalingProps>): void {
    if (!this.Scaling) throw new Error('Scaling needs to be enabled');
    if (!this.TargetGroup) throw new Error('Routing needs to be enabled');
    this.Scaling.scaleOnRequestCount('RequestScaling', {
      ...props,
      requestsPerTarget: 500,
      targetGroup: this.TargetGroup,
    });
  }
}
