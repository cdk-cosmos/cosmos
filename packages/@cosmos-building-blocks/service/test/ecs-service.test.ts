import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { App, Stack } from '@aws-cdk/core';
import { EcsService } from '../src';
import { Cluster, ContainerImage } from '@aws-cdk/aws-ecs';
import { Vpc, SecurityGroup } from '@aws-cdk/aws-ec2';
import { ApplicationListener, ListenerCondition } from '@aws-cdk/aws-elasticloadbalancingv2';

const app = new App();
const stack = new Stack(app, 'Stack');
const vpc = Vpc.fromVpcAttributes(stack, 'Vpc', {
  vpcId: 'Vpc',
  availabilityZones: ['a'],
  isolatedSubnetIds: ['a-sub-1'],
});
const cluster = Cluster.fromClusterAttributes(stack, 'Cluster', {
  clusterName: 'Cluster',
  securityGroups: [SecurityGroup.fromSecurityGroupId(stack, 'SecurityGroup', 'SecurityGroup')],
  vpc,
});
const listener = ApplicationListener.fromApplicationListenerAttributes(stack, 'Listener', {
  listenerArn: 'Listener',
  securityGroup: cluster.connections.securityGroups[0],
});
const listener2 = ApplicationListener.fromApplicationListenerAttributes(stack, 'Listener2', {
  listenerArn: 'Listener2',
  securityGroup: cluster.connections.securityGroups[0],
});
new EcsService(stack, 'EcsService', {
  cluster,
  vpc,
  httpListener: listener,
  httpsListener: listener2,
  httpsRedirect: true,
  containerProps: {
    image: ContainerImage.fromRegistry('Image'),
  },
  routingProps: {
    conditions: [ListenerCondition.pathPatterns(['/path'])],
  },
});

const synth = SynthUtils.toCloudFormation(stack);

test('Ecs Service', () => {
  expect(synth).toHaveResource('AWS::ECS::Service');
  expect(synth).toHaveResource('AWS::ElasticLoadBalancingV2::ListenerRule');
  expect(synth).toHaveResource('AWS::ElasticLoadBalancingV2::TargetGroup');
  expect(synth).toHaveResource('AWS::ECS::TaskDefinition');
  expect(synth).toHaveResource('AWS::Logs::LogGroup');
});

test('Ecs Service Snapshot', () => {
  expect(synth).toMatchSnapshot();
});
