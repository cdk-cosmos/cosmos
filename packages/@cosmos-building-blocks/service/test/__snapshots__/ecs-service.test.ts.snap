// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ecs Service Snapshot 1`] = `
Object {
  "Resources": Object {
    "EcsService0A92EB77": Object {
      "DependsOn": Array [
        "EcsServiceHttpsServiceRuleB6179F15",
      ],
      "Properties": Object {
        "Cluster": "Cluster",
        "DeploymentConfiguration": Object {
          "MaximumPercent": 200,
          "MinimumHealthyPercent": 50,
        },
        "DesiredCount": 1,
        "EnableECSManagedTags": false,
        "HealthCheckGracePeriodSeconds": 60,
        "LaunchType": "EC2",
        "LoadBalancers": Array [
          Object {
            "ContainerName": "Container",
            "ContainerPort": 80,
            "TargetGroupArn": Object {
              "Ref": "EcsServiceServiceTargetGroup6E35E6BD",
            },
          },
        ],
        "PlacementStrategies": Array [
          Object {
            "Field": "attribute:ecs.availability-zone",
            "Type": "spread",
          },
        ],
        "SchedulingStrategy": "REPLICA",
        "TaskDefinition": Object {
          "Ref": "EcsServiceTask59467FF8",
        },
      },
      "Type": "AWS::ECS::Service",
    },
    "EcsServiceHttpsRedirect681F3F00": Object {
      "Properties": Object {
        "Actions": Array [
          Object {
            "RedirectConfig": Object {
              "Port": "443",
              "Protocol": "HTTPS",
              "StatusCode": "HTTP_301",
            },
            "Type": "redirect",
          },
        ],
        "Conditions": Array [
          Object {
            "Field": "path-pattern",
            "PathPatternConfig": Object {
              "Values": Array [
                "/path",
              ],
            },
          },
        ],
        "ListenerArn": "Listener",
        "Priority": 20283,
      },
      "Type": "AWS::ElasticLoadBalancingV2::ListenerRule",
    },
    "EcsServiceHttpsServiceRuleB6179F15": Object {
      "Properties": Object {
        "Actions": Array [
          Object {
            "TargetGroupArn": Object {
              "Ref": "EcsServiceServiceTargetGroup6E35E6BD",
            },
            "Type": "forward",
          },
        ],
        "Conditions": Array [
          Object {
            "Field": "path-pattern",
            "PathPatternConfig": Object {
              "Values": Array [
                "/path",
              ],
            },
          },
        ],
        "ListenerArn": "Listener2",
        "Priority": 20283,
      },
      "Type": "AWS::ElasticLoadBalancingV2::ListenerRule",
    },
    "EcsServiceLogs91D515DC": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "RetentionInDays": 30,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "EcsServiceServiceTargetGroup6E35E6BD": Object {
      "Properties": Object {
        "HealthCheckPath": "/health",
        "Port": 80,
        "Protocol": "HTTP",
        "TargetGroupAttributes": Array [
          Object {
            "Key": "deregistration_delay.timeout_seconds",
            "Value": "0",
          },
        ],
        "TargetType": "instance",
        "VpcId": "Vpc",
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
    "EcsServiceTask59467FF8": Object {
      "Properties": Object {
        "ContainerDefinitions": Array [
          Object {
            "Essential": true,
            "Image": "Image",
            "LogConfiguration": Object {
              "LogDriver": "awslogs",
              "Options": Object {
                "awslogs-group": Object {
                  "Ref": "EcsServiceLogs91D515DC",
                },
                "awslogs-region": Object {
                  "Ref": "AWS::Region",
                },
                "awslogs-stream-prefix": "Container",
              },
            },
            "MemoryReservation": 256,
            "Name": "Container",
            "PortMappings": Array [
              Object {
                "ContainerPort": 80,
                "HostPort": 0,
                "Protocol": "tcp",
              },
            ],
          },
        ],
        "ExecutionRoleArn": Object {
          "Fn::GetAtt": Array [
            "EcsServiceTaskExecutionRole485D054B",
            "Arn",
          ],
        },
        "Family": "StackEcsServiceTask7FDBD854",
        "NetworkMode": "bridge",
        "RequiresCompatibilities": Array [
          "EC2",
        ],
        "TaskRoleArn": Object {
          "Fn::GetAtt": Array [
            "EcsServiceTaskTaskRoleA0BA68C2",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "EcsServiceTaskExecutionRole485D054B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "EcsServiceTaskExecutionRoleDefaultPolicy80218B89": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "EcsServiceLogs91D515DC",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EcsServiceTaskExecutionRoleDefaultPolicy80218B89",
        "Roles": Array [
          Object {
            "Ref": "EcsServiceTaskExecutionRole485D054B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EcsServiceTaskTaskRoleA0BA68C2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "SecurityGroupfromStackSecurityGroup277A41033276865535DAF39B3F": Object {
      "Properties": Object {
        "Description": "Load balancer to target",
        "FromPort": 32768,
        "GroupId": "SecurityGroup",
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": "SecurityGroup",
        "ToPort": 65535,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
  },
}
`;
